lib/astarte_data_access/xandra_utils.ex

```diff
@@ -23,67 +23,120 @@ defmodule Astarte.DataAccess.XandraUtils do
-      opts = opts ++ default_query_opts()
+  def execute_query(conn, statement), do: execute_query(conn, statement, %{}, [])
+
+  def execute_query(conn, statement, opts) when is_list(opts),
+    do: execute_query(conn, statement, %{}, opts)
+
+  def execute_query(conn, statement, params) when is_map(params),
+    do: execute_query(conn, statement, params, [])
+
+  def execute_query(conn, query, params, opts) when is_map(params) and is_list(opts) do
+    opts = Keyword.merge(page_default_opts(), opts)
 
+    with {:ok, prepared} <- prepare_query(conn, query, opts) do
       case Xandra.execute(conn, prepared, params, opts) do
+        {:ok, result} ->
+          {:ok, result}
+
         {:error, %Xandra.Error{} = reason} ->
-          _ = Logger.warn("Database error while retrieving data: #{inspect(reason)}")
+          _ = Logger.warning("Database error while executing: #{inspect(reason)}")
           {:error, :database_error}
 
         {:error, %Xandra.ConnectionError{} = reason} ->
-          _ = Logger.warn("Database connection error while retrieving data: #{inspect(reason)}")
+          _ = Logger.warning("Database connection error while executing: #{inspect(reason)}")
           {:error, :database_connection_error}
-
-        {:ok, %Xandra.Page{} = page} ->
-          {:ok, page}
       end
     end
   end
 
-  defp prepare_query(conn, statement, opts \\ []) do
+  def retrieve_page(conn, query), do: retrieve_page(conn, query, %{}, [])
+
+  def retrieve_page(conn, query, opts) when is_list(opts),
+    do: retrieve_page(conn, query, %{}, opts)
+
+  def retrieve_page(conn, query, params) when is_map(params),
+    do: retrieve_page(conn, query, params, [])
+
+  def retrieve_page(conn, query, params, opts) when is_map(params) and is_list(opts) do
+    with {:ok, %Xandra.Page{} = page} <- execute_query(conn, query, params, opts) do
+      {:ok, page}
+    end
+  end
+
+  defp do_run(fun, keyspace, opts) when is_function(fun) do
+    Xandra.Cluster.run(
+      @cluster_default_name,
+      Keyword.merge([timeout: 60_000], opts),
+      &fun.(&1, keyspace)
+    )
+  end
+
+  defp prepare_query(conn, statement, opts) do
     case Xandra.prepare(conn, statement, opts) do
+      {:ok, %Xandra.Prepared{} = prepared} ->
+        {:ok, prepared}
+
       {:error, %Xandra.Error{} = reason} ->
-        _ = Logger.warn("Database error while preparing query: #{inspect(reason)}")
+        _ = Logger.warning("Database error while preparing query: #{inspect(reason)}")
         {:error, :database_error}
 
       {:error, %Xandra.ConnectionError{} = reason} ->
-        _ = Logger.warn("Database connection error while preparing query: #{inspect(reason)}")
+        _ = Logger.warning("Database connection error while preparing query: #{inspect(reason)}")
         {:error, :database_connection_error}
-
-      {:ok, %Xandra.Prepared{} = prepared} ->
-        {:ok, prepared}
     end
   end
 
-  defp default_query_opts() do
+  defp page_default_opts() do
```

Comment:
These are default values passed to the query, not to the resulting page. I would suggest keeping the old name.